flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,361 @@
+                final SlotSharingGroup slotSharingGroup = slotSharingGroupMap.get(jobVertexID);
+                final int index = getSlotRoundRobinIndex(jobVertexParallel, slotSharingGroup);
+
+                if (!coLocationGroupMap.containsKey(jobVertexID)) {
+                    // For vertices without CoLocationConstraint.
+                    allocateNonCoLocatedVertices(slotSharingGroup, executionVertices, index);
+                } else {
+                    // For vertices with CoLocationConstraint.
+                    allocateCoLocatedVertices(slotSharingGroup, executionVertices, index);
+                }
+            }
+            return executionSlotSharingGroupMap;
+        }
+
+        private void initParalleledExecutionSlotSharingGroupsMap(
+                final LinkedHashMap<JobVertexID, List<SchedulingExecutionVertex>> allVertices) {
+
+            allVertices.entrySet().stream()
+                    .map(
+                            jobVertexExecutionVertices ->
+                                    Tuple2.of(
+                                            slotSharingGroupMap.get(
+                                                    jobVertexExecutionVertices.getKey()),
+                                            jobVertexExecutionVertices.getValue().size()))
+                    .collect(
+                            Collectors.groupingBy(
+                                    tuple -> tuple.f0,
+                                    Collectors.summarizingInt(tuple -> tuple.f1)))
+                    .forEach(
+                            (slotSharingGroup, statistics) -> {
+                                int slotNum = statistics.getMax();
+                                paralleledExecutionSlotSharingGroupsMap.put(
+                                        slotSharingGroup,
+                                        createExecutionSlotSharingGroups(
+                                                slotSharingGroup, slotNum));
+                            });
+        }
+
+        private List<ExecutionSlotSharingGroup> createExecutionSlotSharingGroups(
+                SlotSharingGroup slotSharingGroup, int slotNum) {
+            final List<ExecutionSlotSharingGroup> executionSlotSharingGroups =
+                    new ArrayList<>(slotNum);
+            for (int i = 0; i < slotNum; i++) {
+                final ExecutionSlotSharingGroup executionSlotSharingGroup =
+                        new ExecutionSlotSharingGroup(slotSharingGroup);
+                executionSlotSharingGroups.add(i, executionSlotSharingGroup);
+                LOG.debug(
+                        "Create {}th executionSlotSharingGroup {}.", i, executionSlotSharingGroup);
+            }
+            return executionSlotSharingGroups;
+        }
+
+        private void allocateCoLocatedVertices(
+                SlotSharingGroup slotSharingGroup,
+                List<SchedulingExecutionVertex> executionVertices,
+                int index) {
```

Comment:
This index may be not needed, right?

For max parallelism, using the `executionVertex.getId().getSubtaskIndex()` as the index is enough.

For non-max, I see the index isn't used.